Node/Connection deserialization

maxqchen opened the pull request:
Deserialization for nodes and connections! I only did some basic testing
but everything seems to be good.

Attributes are pretty FUBAR and we definitely need to refactor how they
work in the future, but as far as I know they're not necessary for our
MVP.

Constraints are also still pending.

maxqchen committed:
Node/Connection deserialization

Deserialization for nodes and connections! I only did some basic testing
but everything seems to be good.

Attributes are pretty FUBAR and we definitely need to refactor how they
work in the future, but as far as I know they're not necessary for our
MVP.

Constraints are also still pending.

mtrberzi commented:
We care about a handful of attributes on registers, actually, but all of them are booleans so it shouldn't be too bad to patch that in.

mtrberzi commented on src/main/java/org/manifold/compiler/middle/serialization/SchematicDeserializer.java:
```diff
@@ -46,11 +73,32 @@
+      return attributeMap;
+    }
+
+    // TODO (max): read these, dependent on IR attribute/type overhaul
```
How hard would it be to patch in simple `BooleanValue` deserialization right here?

maxqchen replied:
Shouldn't be hard (Boolean.valueOf into our bool type should do it) but it would be extremely hacky :)

mtrberzi replied:
One thing you could try is, if you have type information about the object whose attributes you're getting, try parsing that type explicitly. 

For example, you're deserializing a node whose type is "register". You look in the Schematic and see that "register" is a node type whose attributes are "foo => bool" and "bar => int". In the attributes section, when you see an attribute whose key is "foo", you deserialize the value as "bool" or throw an exception, and when you see an attribute whose key is "bar", you deserialize the value as "int" or throw an exception.

If you have the type information from earlier, you can use that when you deserialize. Does that make sense?

maxqchen replied:
Yeah. I'll add this when I get back home.

Edit: holy shit. note to self: don't use github email reply on phone

maxqchen committed:
Basic attribute bindings

Quickly threw this together, of course it could be much better...

I think we want some kind of method like TypeValue#createInstance(...)?
Would make this a lot easier

maxqchen committed:
checkstyle pls...

mtrberzi commented on src/main/java/org/manifold/compiler/middle/serialization/SchematicDeserializer.java:
```diff
@@ -46,11 +78,63 @@
+      return attributeMap;
+    }
+
+    // Ideally, some kind of createInstance(String) method on each TypeValue?
```
Yeah, for non-primitive types, value deserialization will become abstract-recursive. But this is fine for now.

mtrberzi commented on src/main/java/org/manifold/compiler/middle/serialization/SchematicDeserializer.java:
```diff
@@ -46,11 +78,63 @@
+      }
+
+      if (type.equals(bool)) {
+        attrValue = BooleanValue.getInstance(Boolean.valueOf(valueString));
```
For the purposes of error detection, what happens if valueString is not something like "true" or "false"? Is there a way to check for this case and throw an exception? (I don't think `valueOf` throws if it can't convert the argument.)

mtrberzi commented on src/test/java/org/manifold/compiler/serialization/TestSerialization.java:
```diff
@@ -131,12 +135,24 @@ public void testDeserialize() throws IOException,
+    assertEquals(digitalIn, andNode.getPort("in1").getType());
+    assertFalse(((BooleanValue) andNode.getAttribute("is_awesome"))
+        .toBoolean());
+    assertTrue(((BooleanValue) andNode2.getAttribute("is_awesome"))
```
Nice, being able to do this check is sufficient for everything I have in code generation for now. MVP hype!

mtrberzi commented on src/main/java/org/manifold/compiler/middle/serialization/SchematicDeserializer.java:
```diff
@@ -46,11 +78,63 @@
+    TypeValue str = sch.getUserDefinedType("String");
+    
+    for (Entry<String, JsonElement> attrEntry : attributeMapJson.entrySet()) {
+      TypeValue type = expectedTypes.get(attrEntry.getKey());
```
What happens if a required attribute is not present? It doesn't look like this can be detected right now, but since all attributes are mandatory, this is important.

maxqchen replied:
You mean an attribute specified in the type but not present in the actual value? That will (should?) be caught in the construction of the type. I'm not sure if this is present for everything but I do recall adding it to Node.

mtrberzi replied:
Oh right, I forgot that this was there. Never mind, that's probably the right place for it then.

mtrberzi commented:
Good work with attributes. I have a couple of nits with exceptional cases, but once those are fixed this is ready to merge.

maxqchen committed:
adding boolean check

mtrberzi commented:
Looks good! I'll try to get this integrated into my backend interface later today. Please merge. :shipit:
